src/cache.rs

```diff
@@ -817,32 +817,42 @@ impl Cache {
     /// Retrieve the id of a source given a name.
     ///
     /// Note that files added via [Self::add_file] are indexed by their full normalized path (cf
-    /// [normalize_path]). When querying file, rather use [Self::id_of_file].
+    /// [normalize_path]).
     pub fn id_of(&self, name: impl AsRef<OsStr>) -> Option<FileId> {
-        self.file_ids.get(name.as_ref()).map(|entry| entry.id)
+        self.id_or_new_timestamp_of(name).ok()
     }
 
-    /// Retrieve the id of a file given a path.
+    /// Try to retrieve the id of a cached source.
     ///
-    /// This function normalizes the given path, search it in the name-id table, and check that the
-    /// stored timestamp is the same as the current timestamp of the file. If normalization or
-    /// metadata retrieval fails, or if the stored entry has no timestamps (it was added as a
-    /// stand-alone source), `None` is returned.
-    pub fn id_of_file(&self, path: impl AsRef<OsStr>) -> io::Result<Option<FileId>> {
-        let normalized = normalize_path(PathBuf::from(path.as_ref()).as_path())?;
-        let timestamp = timestamp(path)?;
-        Ok(self.id_of_file_(normalized, timestamp))
-    }
-
-    /// Retrieve the id of a file given a path. Same as [Self::id_of_file], but assume that the
-    /// given path is already normalized and take the timestamp as a parameter.
-    fn id_of_file_(&self, path: impl AsRef<OsStr>, timestamp: SystemTime) -> Option<FileId> {
-        self.file_ids
-            .get(path.as_ref())
-            .and_then(|entry| match entry.timestamp {
-                Some(ts) if ts == timestamp => Some(entry.id),
-                _ => None,
-            })
+    /// Only returns `Ok` if the source is up-to-date; if the source is stale, returns
+    /// either the new timestamp of the up-to-date file or the error we encountered when
+    /// trying to read it (which most likely means there was no such file).
+    ///
+    /// The main point of this awkward signature is to minimize I/O operations: if we accessed
+    /// the timestamp, keep it around.
```

Comment:
It looks like we are encoding a custom ternary type (`UpToDate(FileId), Stale(Timestamp), Err(err)`). I would suggest to write down this type explicitly and use in the signature. Or at least a type for the first layer of `Result`, because it seems the `Err` branch has nothing do to with an error (or, we're using `Result` as an either). So having a binary type `enum SourceState { UpToDate(FileId), Stale(Timestamp) }` and use `io::Result<SourceState>` would also make sense as well (and might be slightly easier to use with `?` and the like).